### ISSUE TITLE
[indent][meta issue] Problems with the indent rule

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#1824]
### STOP! DO NOT FILE ANY NEW ISSUES ABOUT THE INDENT RULE

Your specific `indent` bug might not have an issue created. **That does not matter**. Don't file a new issue. If you raise an issue we will just close it and reference this issue.
*Please don't waste our time by doing that.*

The rule is broken and telling us different ways in which it is broken doesn't change the fact that it is broken.

For the same reason - please don't comment on this issue with bug reports.

-----

TL;DR - the `indent` rule is broken on TypeScript codebases. Due to bandwidth constraints it will remain broken _**indefinitely**_ unless the community is able to step up to fix it.

# The Problem

Many of you that have used our `indent` extension rule have quickly run into problems with it.

It incorrectly handles many cases that seem trivial, or it handles a block of code in one form, but breaks when you put in a new line. [There are unfortunately a myriad of issues with it](https://github.com/typescript-eslint/typescript-eslint/milestone/1).

So when I first implemented it over a year ago (https://github.com/bradzacher/eslint-plugin-typescript/pull/219), I took a very naive approach to solving the problem - I made TypeScript nodes _look_ like JavaScript nodes, and called existing handlers in the base rule's implementation.

This worked well enough for simple cases, but was quickly discovered to be a terrible solution due to the complexity of TypeScript and the myriad ways in which its constructs can be nested and combined, and also in how users expect each combination to be indented by default.

So I began the process of forking the base rule codebase by copying it over and adding types (#439). I had grand plans to implement every single TypeScript node in this fork, so we could support it fully. This was another naive thought, as it quickly dawned on me how complex the rule's implementation is, and how difficult it is to handle every single variation of each node.

For example, take a simple union type:
```ts
   type T = foo | bar
//          ^^^^^^^^^ expected indent = 0

   type T = foo
//          ^^^ expected indent = 0
     | bar
//   ^^^^^ expected indent = 1

   type T = 
     foo
//   ^^^ expected indent = 1
     | bar
//   ^^^^^ expected indent = 1

   type T =
     | foo
//   ^^^^^ expected indent = 1
     | bar
//   ^^^^^ expected indent = 1
    
   type T = | foo
//          ^^^^^ expected indent = 0
     | bar
//   ^^^^^ expected indent = 1
```

Then there's the notion of configurability for this, which add more complexity.

The base rule has had ~5 years to grow via the community (though it was mostly a few dedicated contributors) - a community that actively uses the rule, and has motivation to use the rule because the problems cause issues for them.

A few years ago, I (like most of the maintainers) made the switch for all of my projects across to prettier. For me, this was spurred by my move to working at facebook, and learning to love the package and everything it does.
This means that I haven't used the rule since then, and I have no personal motivation as a volunteer maintainer to spend the many hours it will take to build this out, and as a maintainer there are many higher-priority issues (such as parsing perf) that need attention.  
As such, the rule has languished, and many issues are open.

# The Solution

The rule is in a broken state. Many users are working around it via ESLint disable comments, and configuring certain nodes as ignored; both of which are poor experiences.

We as maintainers have our hands full maintaining other core infra as well as triaging issues and reviewing PRs, so we don't have the bandwidth to take on this project.

We'd love it if some members of the community could help us out here by submitting some PRs to support TS nodes in the rule fork. Unfortunately, without some champions, the rule will not progress forward or be fixed.

Not to discourage anyone, but as a forewarning: there are some 2,100 lines of code already, and that just constitutes the typechecked fork of the base rule. The codebase contains some rather complex algorithms, and has a lot of moving parts. It will be no small undertaking to understand what it's doing, and how it's doing it.

**Comment by jeffvandyke** [comment]
This is a pretty monster rule, and naturally quite important in the ecosystem. Sounds like a formidable but exciting challenge.

Those referencing issues are pretty useful for documenting some user-expected behavior; I found my specific complaint in #455.

At the risk of sounding foolish, @bradzacher,  is there any other relevant work or notes you've uncovered in your research that might help for understanding this rule? (besides the linked PR #439)

**Comment by bradzacher** [comment]
unfortunately I don't have a _whole_ lot of information - it was a year ago that I looked into the codebase. There's little documentation apart from the sparse comments in the code.

I know that it tracks the indent based on ranges, not per node. This is done because a node can have multiple indentation requirements (eg my example above).

I know it uses a red-black tree to efficiently track and recall this information, though I never spent enough time to learn about how a red-black tree actually works (idk if you really need to know how it works to make the rule work).

Beyond that I don't really know. I spent a little bit of time toying with the rule to figure out how it worked, but I didn't get much further than that due to time constraints.

> Sounds like a formidable but exciting challenge

Definitely - were I not a volunteer maintainer with a wife and a full time job, I would gladly spend days diving deep into a complete rewrite of the rule. Learning and understanding this implementation, doing the same for prettier's indentation algorithm, and using that knowledge to build a _strict_, configurable, and performant rule would be such an interesting problem to solve.

**Comment by jeffvandyke** [comment]
I think it would be good for the ecosystem if `@typescript-eslint/indent` says mostly in sync with `indent` from ESLint. Other than refactoring and schema additions done here, I think they're fairly close right now, so any applicable work done here should go back to ESLint (if it fixes bugs for JavaScript).

Another related question, since this rule worries about differences between TypeScript and JavaScript's Node types: How does this play in to the effort to refine a superset of JavaScript AST nodes that can accommodate TypeScript + Babel setup for ESLint? I only remember reading that it was an ongoing effort. Is it possible and desired that ESLint's `indent` rule will one day handle TypeScript parsed by Babel?

**Comment by bradzacher** [comment]
As part of the hypothetical rewrite, I'd upstream it to eslint as a new rule `indent-strict` (or something), with the intention of them deprecating `indent`.

The `indent` rule is in a really ***weak*** state right now because it has maintained backwards compatibility for a long time, so there are things it doesn't enforce at all.

Eg this is valid according to the rule, [because backwards compatibility](https://github.com/typescript-eslint/typescript-eslint/blob/cc70e4fbadd0b15fd6af913a2e1e2ddd346fa558/packages/eslint-plugin/src/rules/indent-new-do-not-use/index.ts#L912-L932)
```ts
const x = 1 &&
                           2 &&
3;
```
[repl](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IFwiaW5kZW50XCI6IFwiZXJyb3JcIiAqL1xuY29uc3QgeCA9IDEgJiZcbiAgICAgICAgICAgICAgICAgICAgICAgICAgIDIgJiZcbjM7XG5cbmZ1bmN0aW9uIGZvbygpIHtcbmJhcjsgLy8gZXhwZWN0ZWQgZXJyb3IgLSA0IHNwYWNlc1xufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OSwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

---

There's little difference between the ASTs provided by `typescript-estree` and `babel-eslint`. There's a few minor inconsistencies, but IIRC most of those are things to be fixed on the babel side, and don't have a huge impact on lint rules.

**Comment by JakubKoralewski** [comment]
Is there a way to enforce tabs, and throw errors where spaces are used for indentation, but that would ignore errors regarding how many levels of indentation there are in all contexts, for all kinds of expressions?

**Comment by bradzacher** [comment]
https://eslint.org/docs/rules/no-mixed-spaces-and-tabs

**Comment by JakubKoralewski** [comment]
This will throw an error if I mix tabs and spaces but won't do anything if
someone uses just spaces, as far as I can see

**Comment by garyo** [comment]
@plohoj that is useful info re: AST Explorer. Could it help with this mis-indentation situation?
```ts
export const foo = functions.https.onCall(async (data: {width: number;
                                                        height: number;
                                                        type: 'video'|'still';},
  context) => {
  setLogStackdriverMode()
  ...
}
```
where `typescript-eslint` insists that the arg `context` should be indented two spaces, rather than lining up under the first arg, `data`?

**Comment by plohoj** [comment]
@garyo I have not worked with AST. Perhaps my decision will not be correct, but you can turn off the check in this way.
```js
'@typescript-eslint/indent': ['error', 4, {
    ignoredNodes: [
        'ArrowFunctionExpression[params]',
        'ArrowFunctionExpression[params] TSPropertySignature'
    ],
}],
```
As you can see, this only partially solves the problem, because the nested blocks are not aligned with the previous ones. In this case, it is easier to disable checking for everything that is embedded inside this block.
```js
'@typescript-eslint/indent': ['error', 4, {
    ignoredNodes: [
        'ArrowFunctionExpression Identifier',
        'ArrowFunctionExpression BlockStatement',
    ]
}],
```
You can choose the selector that best suits your situation.

**Comment by aristocr2t** [comment]
I have a problem using the generics method:
```typescript
class SomeClass {
	test(): Observable<any> {
		return this.http
			.get<Test[]>(`test`, {
			headers: { authorization: 'some value' },
		});
	}
}
```
should be:
```typescript
class SomeClass {
	test(): Observable<any> {
		return this.http
			.get<Test[]>(`test`, {
				headers: { authorization: 'some value' },
			});
	}
}
```
but when I remove generics from method - issue disappear:
```typescript
class SomeClass {
	test(): Observable<any> {
		return this.http
			.get(`test`, {
				headers: { authorization: 'some value' },
			});
	}
}
```

**Comment by aristocr2t** [comment]
when I write a new "type" expression inside a function, next lines take indent size from "type" expression:
```typescript
function test(): void {
		type Test = 'test'; // 2 tabs, but needs 1 tab

		const test = 'test'; // also 2 tabs, because it takes from "type" expression
}
```
should be:
```typescript
function test(): void {
	type Test = 'test'; // 1 tab

	const test = 'test'; // 1 tab
}
```

**Comment by thany** [comment]
Here's my 3 cents:

Example 1: Incorrectly marked as incorrect, but correct for the config I gave it (all defaults for the `indent` rule):
```typescript
const fn = (): string[] => ['foo', 'bar']
  .filter(Boolean)
  .map<string>(child => {
    console.log('It wants to unindent here');
    return child;
  });
```

Example 2: This is *not* marked as incorrect:
```typescript
const fn = (): string[] => ['foo', 'bar']
  .filter(Boolean)
  .map(child => {  // Missing generic type now!
    console.log('It DOESN'T want to unindent here now');
    return child;
  });
```

Example 3: It wants me to make it into this atrocity:
```typescript
const fn = (): string[] => ['foo', 'bar']
  .filter(Boolean)
  .map<string>(child => {
  console.log('It thinks this is okay...');
  return child;
});
```

So it has to do with interpreting typescript stuff specifically. Worse is that example 3, leaving out the generic type again, is marked as incorrect again, and it wants me to make it into example 2.

**Comment by TobiasWehrum** [comment]
Just a note for those who (like me) have the same problems with generic function calls as described in the last post - I almost overlooked the workaround linked to in https://github.com/typescript-eslint/typescript-eslint/issues/1824#issuecomment-678640058, and this seems to work for now:

```
'@typescript-eslint/indent': ['error', 2, {
  ignoredNodes: ['TSTypeParameterInstantiation']
}]
```

**Comment by sagrawal31** [comment]
> Just a note for those who (like me) have the same problems with generic function calls as described in the last post - I almost overlooked the workaround linked to in [#1824 (comment)](https://github.com/typescript-eslint/typescript-eslint/issues/1824#issuecomment-678640058), and this seems to work for now:
> 
> ```js
> '@typescript-eslint/indent': ['error', 2, {
>   ignoredNodes: ['TSTypeParameterInstantiation']
> }]
> ```

This solution not working for me-

![image](https://user-images.githubusercontent.com/1804514/134331123-4f071c75-2e32-44b3-8bee-acac2a2d2e93.png)

Configuration-

```
indent: 'off',
'@typescript-eslint/indent': ['error', 4, {
    'FunctionDeclaration': {
        'body': 1,
        'parameters': 'first',
    },
    'FunctionExpression': {
        'body': 1,
        'parameters': 0,
    },
    ignoredNodes: ['TSTypeParameterInstantiation']
}],
```

Dependencies-

```
"@typescript-eslint/eslint-plugin": "^4.31.2",
"@typescript-eslint/parser": "^4.31.2",
"eslint": "^7.32.0",
```

**Comment by TobiasWehrum** [comment]
> 
> 
> > Just a note for those who (like me) have the same problems with generic function calls as described in the last post - I almost overlooked the workaround linked to in [#1824 (comment)](https://github.com/typescript-eslint/typescript-eslint/issues/1824#issuecomment-678640058), and this seems to work for now:
> > ```js
> > '@typescript-eslint/indent': ['error', 2, {
> >   ignoredNodes: ['TSTypeParameterInstantiation']
> > }]
> > ```
> 
> This solution not working for me-
> 
> ![image](https://user-images.githubusercontent.com/1804514/134331123-4f071c75-2e32-44b3-8bee-acac2a2d2e93.png)

That solution is specifically for the issue with functions with generic type parameters described here: https://github.com/typescript-eslint/typescript-eslint/issues/1824#issuecomment-810953933

Unfortunately I know next to nothing about this topic, so I can't help you adapt it.

PS: It's helpful if you also post your code in a code block so anyone who wants to help you can quickly try to reproduce it themselves.

**Comment by sagrawal31** [comment]
Thanks, @TobiasWehrum for replying. Yes, I have tried different PnC for those values (that includes the config you mentioned of using `parameters` to `first` or 1). That solves the error on the `constructor` but starts throwing for other methods-

![image](https://user-images.githubusercontent.com/1804514/134350375-310b8965-cabf-4c7e-a8bb-67eaa52fa960.png)

(*ESLint: Expected indentation of 12 spaces but found 8.(@typescript-eslint/indent)*)

----


> Maybe I'm misunderstanding it, but I'd have thought that putting it to 0 wants it to like this:

Yes, you are right.

**Comment by sagrawal31** [comment]
And a weird thing is that (which I'm not able to understand) if I remove the `: Promise<IResponse>` from the `create` method return type then the error starts to come on the first line-

![image](https://user-images.githubusercontent.com/1804514/134354989-51d7fa34-d02c-4cde-88ea-75c83c53788a.png)

(*ESLint: Expected indentation of 4 spaces but found 8.(@typescript-eslint/indent)*)

**Comment by TobiasWehrum** [comment]
@sagrawal31 Ah, now that actually looks like a bug!

Just like the workaround in my previous comment it's not ideal, but try if `"ignoredNodes": ["Decorator"]` works for you.

**Comment by sagrawal31** [comment]
Yes, @TobiasWehrum. It's preventing the error for any parameter which is having a decorator. Here is the expected behaviour that is happening correct now after using `ignoreNodes`-

![image](https://user-images.githubusercontent.com/1804514/134363661-e1504f7d-37da-494c-93a9-13f5e07538b3.png)


Thanks for your help. I'll for now use this. But should we raise another bug for this?

**Comment by TobiasWehrum** [comment]
@sagrawal31
Yeah, ignoring stuff will sadly lead to some errors not showing. I don't have a workaround for that, and I'm not sure one exists until this issue here is properly resolved somehow, which I wouldn't expect any day soon judging by the first post.

Someone from this project would probably know better than me, but it seems to me that *this* issue is the big catch-all issue for all indent problems, so I'd assume no other bug issue should be raised.

**Comment by egorovli** [comment]
I've sort of fixed this with a band-aid for decorated class properties and function params:

```jsonc
{
  // ...
  "rules": {
    // ...
    "indent": "off",
    "@typescript-eslint/indent": [
      "error",
      2,
      {
        "ignoredNodes": [
          "FunctionExpression > .params[decorators.length > 0]",
          "FunctionExpression > .params > :matches(Decorator, :not(:first-child))",
          "ClassBody.body > PropertyDefinition[decorators.length > 0] > .key"
        ]
      }
    ]
  }
}
```

Now you don't need to ignore these cases on a per-file basis:

```diff
@Controller('/something')
export class SomeController {
  @Post()
  async doSomething (
+   @Body() body: DoSomethingBodyDto,
-     @Body() body: DoSomethingBodyDto,
    @Query() query: DoSomethingQueryDto
  ): Promise<unknown> {
    // ...
  }
}
```

```diff
export class Entity {
  @PrimaryKey()
+ id!: string
-   id!: string

  @Property()
+ amount!: number
-   amount!: number
}
```

**Comment by thany** [comment]
Why is every comment being marked as off-topic? As far as I can tell, they are very much on-topic...

**Comment by bradzacher** [comment]
We don't hugely want this to be a support thread for people to trade hacks about ways to ignore things to satisfy the rule.
A support forum like stackoverflow would be a better place to ask for, and discuss that, and it would be more searchable there as well.

Ultimately people asking for help or posting hacks just creates noise for us maintainers - we can't help you ourselves. But because we want to know if someone actually wants to work on improving the rule, we need to keep subscribed to this issue.

So collapsing all the comments is intended to discourage people from using this as a dumping ground to help keep the issue high signal for us maintainers.

**Comment by nickofthyme** [comment]
That's understandable wanting to dampen the noise of workarounds to this issue. However, you are requesting people not open separate issues for specific cases with the `indent` rule. That's fine but stackoverflow is not the solution, that is where people go to solve solutions to their problems assuming the code is working correctly, which respectfully, is not **yet** the case.

Imagine react had a feature that didn't function correctly, posting an issue on GH would be the logical response, not going to stack overflow, because people on SO are not going to fix the issue. And since you are requesting no issues be opened around specific cases that leaves people to comment on this issue.

Would you be open to enabling the GH discussions tab and directing people there to discuss interim solutions to specific cases? This would also give you all an idea of issues sill needing fixes without have to search SO.

**Comment by nickofthyme** [comment]
Ok that's fine, what about a separate issue to serve as a restrictive discussion? Then you can lock this issue to allow only maintainers to comment. Just a thought, otherwise I think people will continue to comment here.

**Comment by DarkRoku12** [comment]
This does not solve the rule, but at least I'd like to share my solution I've working in this weekend.

For my purpouses, this does great indentation, _at least for my codebase_, we could extend the rule to add more 'cases', but I get most of the normal code are covered.

Note that this rule have no other configurations aside for being turned on/off, and it is a little bit opinionated, it can be tuned to expose some configuration to ESLint to use indentation level (1,2,3,4… spaces or tabs), so it does assume you'll use 2 indentation spaces (not tabs).

As you may notice, I'm short of time, but I could expend more time working on it to make it more customizable, but maybe at the end of this month. That why I'm creating a gist rather than a repo.

https://gist.github.com/DarkRoku12/247defa56ac6475043a42145d51ab013

The gist file itself is indented with that rule (no other indentation rules are enabled).

It should support both JavaScript and TypeScript, since my project is in TypeScript.

Hope I can help somebody with this.

**Comment by gajus** [comment]
<img width="547" alt="Screen Shot 2021-12-22 at 10 20 33 AM" src="https://user-images.githubusercontent.com/973543/147123328-e152c580-88f4-4c51-bf68-c31704c2b391.png">

@DarkRoku12 Doesn't work with JSX, but otherwise a vast improvement.

Would you want to work on this if I published it under https://github.com/gajus/eslint-plugin-canonical ?

**Comment by DarkRoku12** [comment]
> <img alt="Screen Shot 2021-12-22 at 10 20 33 AM" width="547" src="https://user-images.githubusercontent.com/973543/147123328-e152c580-88f4-4c51-bf68-c31704c2b391.png">
> 
> @DarkRoku12 Doesn't work with JSX (puts everything on the same line), but otherwise a vast improvement.
> 
> And if you try to to use it together with `react/jsx-indent`, the rules are clashing.
> 
> Would you want to work on this if I published it under https://github.com/gajus/eslint-plugin-canonical ?

Yes, I can work on that, I got more available time now.
I'll see if I could add an option for ignoring JSX nodes.

**Comment by karikera** [comment]
![image](https://user-images.githubusercontent.com/6016498/162557756-c396e3c3-dc4a-4f98-a344-c33342f697eb.png)

are there any documents about type names of `ignoredNodes`?
I want to ignore the union-type issue. but I cannot find the way.

**Comment by ben12** [comment]
> 

Workaround to ignore this issue:
```json
        "@typescript-eslint/indent": [
          "error",
          2,
          {
            "ignoredNodes": [
              "PropertyDefinition[decorators]",
              "TSUnionType"
            ]
          }
        ],
```

Any other "ignoredNodes" to add ?

**Comment by Kreozot** [comment]
> Any other "ignoredNodes" to add ?

TSTypeParameterInstantiation
TSIntersectionType

**Comment by sacru2red** [comment]
anyone knows selecting decorators at ignoredNodes?

**Comment by IMalyugin** [comment]
One of the common issues - class method parameter decorators:
```
export class Foo {
  bar(
  @Request() req: any,   <--
    @Body() body: any,
    @Query() params: any,
  ) {
    ...
  }
```

Can be fixed by adding `"FunctionExpression[params]:has(Identifier[decorators])"` to the `ignoredNodes` list.

ignoredNodes accept a custom AST selector, and selectors are quite powerful. The given selector disables rule for the entire  function params section if any of them contains a decorator. If that is too much, another :not() pseudo selector may be applied to disable the behavior for any of the nested nodes.

**Comment by mb21** [comment]
@IMalyugin very cool, thanks! This indeed doesn't error anymore on decorators:

```json
"@typescript-eslint/indent": ["error", 2, {"ignoredNodes": ["FunctionExpression[params]:has(Identifier[decorators])"]}],
```

But weirdly, it now errors on:

```js
switch (foo) {
  case 'bar':
```

saying `case should be indented 2 spaces less`... even though there isn't any decorator on that function at all.

**Comment by IMalyugin** [comment]
@mb21 you were probably using indent provided by another plugin, for example airbnb. If so, you have to "add new ignored nodes" to it, not override it.

My setup is this:
```
const { rules: baseStyleRules } = require('eslint-config-airbnb-base/rules/style');
...
'@typescript-eslint/indent': [2, 2, {
  ...baseStyleRules.indent[2],
  ignoredNodes: [
    ...baseStyleRules.indent[2].ignoredNodes,
    'PropertyDefinition[decorators]',
    'TSUnionType',
    // 'FunctionExpression[params]:has(Identifier[decorators])',
    'FunctionExpression[params]:has(Identifier[decorators])',
    // FIXME: Recommended to add these two as well, see if there is an error
    // 'TSTypeParameterInstantiation',
    // 'TSIntersectionType',
  ]
}]
```

Note that eslintrc is 'js', not json, otherwise I would have to copy the entire extended rule.

**Comment by mb21** [comment]
@IMalyugin no, we have a `.eslint.json` file with:

```json
{
  "parser": "@typescript-eslint/parser",
  "extends": [
    "standard",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended"
  ],

...
  "rules": {
    "indent": "off",
    "@typescript-eslint/indent": ["error", 2, {"ignoredNodes": ["FunctionExpression[params]:has(Identifier[decorators])"]}],
```

and the error is clearly from typescript-eslint:

```
error  Expected indentation of 2 spaces but found 4  @typescript-eslint/indent
```

🤷

**Comment by sacru2red** [comment]
@mb21

super cool!
Thank you

**Comment by IMalyugin** [comment]
@mb21 as I said earlier, you probably had some other default, that got overridden by adding custom `@typescript-eslint/indent`.

According to the source code of typescript-eslint rules, the default value for `@typescript-eslint/indent` is the following:
```
'@typescript-eslint/indent': ['2', 4, {
   // typescript docs indent the case from the switch
   // https://www.typescriptlang.org/docs/handbook/release-notes/typescript-1-8.html#example-4
   SwitchCase: 1,
   flatTernaryExpressions: false,
   ignoredNodes: [],
}],
```
Came from here: https://github.com/typescript-eslint/typescript-eslint/issues/608

So if you want to merge the default behavior, with decorators fix, you need the following configuration:

```json
{
  "parser": "@typescript-eslint/parser",
  "extends": [
    "standard",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended"
  ],
...
  "rules": {
    "indent": "off",
    "@typescript-eslint/indent": ["error", 2, {
      "ignoredNodes": [
        "PropertyDefinition[decorators]",
        "TSUnionType",
        "FunctionExpression[params]:has(Identifier[decorators])"
      ]
    }],

**Comment by mb21** [comment]
@IMalyugin ah, I see... thanks a lot! working great 🥳

**Comment by witcher112** [comment]
As a temporary solution for indent rule being in a broken state, I decided to create a plugin that brings the built-in TypeScript formatter to ESLint (which I find to be working great!).

You can access it through npm: https://www.npmjs.com/package/eslint-plugin-typescript-formatter

I hope this can help someone, cheers!

**Comment by SalahAdDin** [comment]
> As a temporary solution for indent rule being in a broken state, I decided to create a plugin that brings the built-in TypeScript formatter to ESLint (which I find to be working great!).
> 
> You can access it through npm: https://www.npmjs.com/package/eslint-plugin-typescript-formatter
> 
> I hope this can help someone, cheers!

We use **Prettier**.

**Comment by binury** [comment]
This is not a discussion thread support chat. You are _emailing **everyone** watching this issue_ when you reply.

**Comment by 500-internal-server-error** [comment]
Hello, given that [this page](https://typescript-eslint.io/linting/troubleshooting/formatting/) mentions the following:

> Per ESLint's 2020 Changes to Rule Policies blog post, ESLint itself has moved away from stylistic rules, including formatting rules:
>
> > Stylistic rules are frozen - we won't be adding any more options to stylistic rules. We've learned that there's no way to satisfy everyone's personal preferences, and most of the rules already have a lot of difficult-to-understand options. Stylistic rules are those related to spacing, conventions, and generally anything that does not highlight an error or a better way to do something.
>
> We support the ESLint team's decision and backing logic to move away from formatting and stylistic rules. With the exception of bug fixes, no new formatting- or stylistic-related pull requests will be accepted into typescript-eslint.

Is fixing this rule still being worked on / actively supported?

Thanks in advance.

**Comment by bradzacher** [comment]
Given this issue has been open for 3 years with no community member pitching in (which is understandable!) - it's clear that this rule is not going to be fixed here.

Following on from https://github.com/eslint/eslint/issues/17522
ESLint is moving to deprecate and remove formatting rules - meaning rules like `indent` will be moved to the community domain. We will follow suit when they do as well and will similarly deprecate our formatting rules.

With those two signals - I am going to lock this issue to prevent any further comments, and we shall not be accepting any PRs to fix the rule. I'll leave this open so it shows up in search, but consider it functionally closed.

If you need further assistance with the rule - please consider posting on stackoverflow or similar.
